dev/buildtool/source_code_manager.py

```diff
@@ -0,0 +1,294 @@
+      logging.error(message)
+      raise ValueError(message)
+
+    logging.info('Pulling %s to %s', repository.url, git_dir)
+    ensure_dir_exists(os.path.dirname(git_dir))
+
+    if git_branch:
+      self.git.clone_repository_to_path(
+          repository.url, git_dir,
+          branch=git_branch, default_branch=default_branch)
+    else:
+      if bom_version:
+        bom = self.bom_from_version(bom_version)
+      else:
+        bom = self.bom_from_path(bom_path)
+      self.pull_source_from_bom(repository.name, git_dir, bom)
+
+  def pull_source_from_bom(self, repository_name, git_dir, bom):
+    """Pull the source for a particular repository name from a bom."""
+    if repository_name in ('monitoring-daemon', 'monitoring-third-party'):
+      repository_name = 'spinnaker-monitoring'
+
+    spec = bom['services'][repository_name]
+    commit = spec['commit']
+    version = spec['version']
+
+    git_tag = 'version-{}'.format(version[:version.index('-')])
+
+    # We're creating a local repo containing only the commit and tag.
+    # The reason for this is because nebula gets confused by our tags
+    # and wants to only use the latest Netflix tag, which is wrong.
+    # If we could control nebula, we wouldnt need to wipe the repo and
+    # add a new one, but we'd still need to add the new tag to ensure it
+    # exists because the bom may not have been pushed.
+    if os.path.exists(git_dir):
+      return None
+
+    base_url = bom['artifactSources']['gitPrefix']
+    origin_url = spec.get(
+        'sourceRepository',
+        '{base}/{name}'.format(base=base_url, name=repository_name))
+    branch = spec.get(
+        'sourceBranch',
+        bom['artifactSources'].get('gitBranch', '(unknown)'))
+
+    self.__git.clone_repository_to_path(origin_url, git_dir, commit=commit)
+    self.__git.reinit_local_repository_with_tag(
+        git_dir, git_tag,
+        'Repository snapshot from BOM\n'
+        '\ncommit={commit}\nurl={url}\nbranch={branch}'.format(
+            commit=commit, url=origin_url, branch=branch))
+
+  def foreach_source_repository(
+      self, call_function, *posargs, **kwargs):
+    """Call the function on each of the SourceRepository instances."""
+    use_threadpool = kwargs.pop('use_threadpool', False)
+
+    worker = RepositoryWorker(call_function, *posargs, **kwargs)
+    all_repos = self.__source_repositories
+    num_threads = min(self.__max_threads, len(all_repos))
+    if num_threads > 1:
+      if True or use_threadpool:
+        pool = ThreadPool(num_threads)
+      else:
+        pool = Pool(num_threads, maxtasksperchild=1)
+      logging.info('Mapping with %s %d/%s',
+                   pool.__class__,
+                   len(all_repos.keys()), all_repos.keys())
+      try:
+        raw_list = pool.map(worker, all_repos.values())
+        result = {name: value for name, value in raw_list}
+      except Exception:
+        logging.error('Map caught exception')
```

Comment:
Will we have more contextual information about the failure other than this, or can we dump the exception or exception, task pair for more traceability?

Reply:
In practice the exeption was already logged when it was thrown so logging the details here is redundant. This message here is confirmation that we unwound this far afterward.

Reply:
Ok, makes sense.